### PR TITLE
test(cluster): Add basic cluster mode pytests.

### PR DESCRIPTION
**Pull request by chakaz**
When reviewing, please note that:
1. I don't speak Python
2. It's my first time modifying pytests
So basically - I don't know what I'm doing, so do not assume any competence :)

### CODE REVIEW

File: tests/dragonfly/cluster_test.py [hunk 1]
```diff
@@ -142,3 +148,121 @@ async def test_cluster_nodes(async_client):
...
+    ]
+
+    df_local_factory.start_all(masters)
+
+    c_masters = [aioredis.Redis(port=master.port) for master in masters]
+    c_masters_admin = [aioredis.Redis(port=master.admin_port) for master in masters]
+
+    master_ids = []
+
+    for i in range(len(masters)):
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:167]:
```
   master_ids = await asyncio.gather(*(admin_client.execute_command("DFLYCLUSTER MYID")
                           for c_admin in c_masters_admin))
```

File: tests/dragonfly/cluster_test.py [hunk 2]
```diff
@@ -8,6 +8,12 @@
 BASE_PORT = 30001
 
 
+async def push_config(config, admin_connections):
+    print("Pushing config ", config)
+    for c in admin_connections:
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:13]: use asyncio.gather here as well

File: tests/dragonfly/cluster_test.py [hunk 3]
```diff
@@ -142,3 +148,121 @@ async def test_cluster_nodes(async_client):
     assert info['last_ping_sent'] == '0'
     assert info['slots'] == [['0', '16383']]
     assert info['master_id'] == "-"
+
+
+@dfly_args({"proactor_threads": 4, "cluster_mode": "yes"})
+async def test_cluster(df_local_factory):
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:154]: lets be more specific in the test name. What are we testing in this test case?
Also adding a comment above the test for the steps in the test is good.

File: tests/dragonfly/cluster_test.py (hunk 1 above)
Comment (chakaz) [reply tests/dragonfly/cluster_test.py:167]: Just FYI that I needed to `decode()` each ID, which slightly made this statement more complex, but I followed through with your comment

File: tests/dragonfly/cluster_test.py [hunk 4]
```diff
@@ -150,8 +151,13 @@ async def test_cluster_nodes(async_client):
...
+# Test that slot ownership changes correctly with config changes:
+# Add a key to master0, then move the slot ownership to master1 and see that they both behave as
+# intended.
+# Also add keys to each of them that are *not* moved, and see that they are unaffected by the move.
 @dfly_args({"proactor_threads": 4, "cluster_mode": "yes"})
-async def test_cluster(df_local_factory):
+async def test_cluster_slot_ownership_changes(df_local_factory):
+    # Start and configure cluster with 2 masters
     masters = [
         df_local_factory.create(port=BASE_PORT+i+1, admin_port=BASE_PORT+i+1001)
```
[the remark this answers is not in the document]
Comment (chakaz) [reply tests/dragonfly/cluster_test.py]: I guess I don't :) Oops

File: tests/dragonfly/cluster_test.py [hunk 5]
```diff
@@ -162,12 +168,11 @@ async def test_cluster(df_local_factory):
...
 
-    master_ids = []
-
-    for i in range(len(masters)):
-        id = (await c_masters_admin[i].execute_command("DFLYCLUSTER MYID")).decode()
-        print(f'Master {i} id is {id}')
-        master_ids.append(id)
+    master_ids = [
+        id.decode()
+        for id in await asyncio.gather(*(c_admin.execute_command("DFLYCLUSTER MYID")
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:173]: you can write a small function calling execute_command("DFLYCLUSTER MYID") and returning the decoded result. 
asyncio.gather with this function

Comment (chakaz) [reply tests/dragonfly/cluster_test.py:173]: I'm far for an expert on Python `async`, so I'm probably wrong. But I was under the impression that `gather` works on futures / async stuff, and so I'll need to pass in something of that family.
Now, `decode()` is a regular (sync) function, so how would that work?

```python
(async execute_command("DFLYCLUSTER MYID")).decode()
```
Is no longer async right?

File: tests/dragonfly/cluster_test.py [hunk 6]
```diff
@@ -142,3 +149,134 @@ async def test_cluster_nodes(async_client):
...
+
+
+# Test that slot ownership changes correctly with config changes:
+# Add a key to master0, then move the slot ownership to master1 and see that they both behave as
+# intended.
+# Also add keys to each of them that are *not* moved, and see that they are unaffected by the move.
+@dfly_args({"proactor_threads": 4, "cluster_mode": "yes"})
+async def test_cluster_slot_ownership_changes(df_local_factory):
+    # Start and configure cluster with 2 masters
+    masters = [
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:161]: nit: there is no master and replica in this test.
I would just call them nodes

File: tests/dragonfly/cluster_test.py [hunk 7]
```diff
@@ -142,3 +149,134 @@ async def test_cluster_nodes(async_client):
...
+        }}
+      ]
+    """
+
+    await push_config(config, c_masters_admin)
+
+    # Slot for "KEY1" is 5259
+
+    # Insert a key that should stay in master0
+    assert await c_masters[0].execute_command("SET", "KEY0", "value")
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py]: nit: client.set("key","value")

File: tests/dragonfly/cluster_test.py (hunk 5 above)
Comment (adiholden) [reply tests/dragonfly/cluster_test.py:175]: This does not work?
```
async def get_my_id(client):
  id = await execute_command("DFLYCLUSTER MYID")
  return id.decode()
```

File: tests/dragonfly/cluster_test.py [hunk 8]
```diff
@@ -142,3 +149,136 @@ async def test_cluster_nodes(async_client):
     assert info['last_ping_sent'] == '0'
     assert info['slots'] == [['0', '16383']]
     assert info['master_id'] == "-"
+
+
+""" Test that slot ownership changes correctly with config changes.
```
Comment (adiholden) [inline tests/dragonfly/cluster_test.py:154]: start your comment line below """

File: tests/dragonfly/cluster_test.py (hunk 5 above)
Comment (chakaz) [reply tests/dragonfly/cluster_test.py]: It worked! Thanks :)

**Review by adiholden — APPROVED** [review]

**Pull request merged**